src/View/Helper/PaginatorHelper.php

```diff
@@ -1228,6 +1233,57 @@ public function limitControl(array $limits = [], ?int $default = null, array $op
         return $out;
     }
 
+    /**
+     * Prepare and filter limit options for limitControl.
+     *
+     * Handles generating limits from steps, applying defaults, and filtering by maxLimit.
+     *
+     * @param array<string, string> $limits Explicit limit options
+     * @param int|null $steps If provided, generates limits in multiples of this value
+     * @return array<int|string, string> Prepared limit options
+     */
+    protected function prepareLimitOptions(array $limits, ?int $steps): array
+    {
+        // Generate limits based on steps if provided
+        if ($steps !== null) {
+            if ($limits !== []) {
+                throw new InvalidArgumentException(
+                    'Cannot use both `steps` option and explicit `$limits` array. ' .
+                    'Use one or the other.',
+                );
+            }
+
+            $maxLimit = $this->param('maxLimit');
+            $upperLimit = $maxLimit ?? 100;
+            $limits = [];
+            for ($i = $steps; $i <= $upperLimit; $i += $steps) {
+                $limits[$i] = (string)$i;
+            }
+
+            return $limits;
+        }
+
+        // Apply default limits if none provided
+        $limits = $limits ?: [
+            '20' => '20',
+            '50' => '50',
+            '100' => '100',
+        ];
+
+        // Filter out limits that exceed maxLimit
+        $maxLimit = $this->param('maxLimit');
+        if ($maxLimit !== null) {
```

Comment:
Isn't the filtering done inside this `if` block redundant if `$limits` was generated based on `$steps`, which already max limit as the upper limit?  So check should probably be `if ($steps !== null && $maxLimit !==null)`

Reply:
I've refactored prepareLimitOptions() to use an early return when steps is provided (line 1263), so the filtering logic is never reached in that case. Now the two code paths are clearly separated:
  - steps path: generates limits up to maxLimit, returns immediately
  - explicit/default limits path: applies maxLimit filtering

  No redundant filtering occurs.